### ISSUE TITLE
Initializing primitive from literal with cast

### ISSUE THREAD

**Issue by gnzlbg** [rust-lang/rust-clippy#3842]
I've seen code like this:

```rust
fn main() {
    let _ = 100 as usize;
    let _ = 100 as f64;
}
```

where clippy currently errors with

```
warning: casting i32 to f64 may become silently lossy if types change
 --> src/main.rs:3:13
  |
3 |     let _ = 100 as f64;
  |             ^^^^^^^^^^ help: try: `f64::from(100)`
  |
  = note: #[warn(clippy::cast_lossless)] on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#cast_lossless
```

I think we should at least suggest if people didn't meant to write:

```rust
fn main() {
    let _ = 100_usize;
    let _ = 100_f64;
}
```

instead. Semantically, `100_i32 as f64` and `100_f64` are not the exact same thing, but in practice..

**Comment by rink1969** [comment]
I'd like to try this issue.
But I'm not sure what is expected output?
Is something like this OK?
```
warning: casting integer literal to f64 is unnecessary
 --> examples/clippytest.rs:3:13
  |
3 |     let y = 100 as f64;
  |             ^^^^^^^^^^ help: try: `100_f64`
  |
  = note: #[warn(clippy::unnecessary_cast)] on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_cast
````

**Comment by gnzlbg** [comment]
cc @oli-obk I'm not sure whether suggesting `100_f64` only is ok, or whether both should be suggested.

**Comment by oli-obk** [comment]
Yes I think we should be extending the `cast_lossless` lint to suggest both solutions. This doesn't require a new lint.

**Comment by rink1969** [comment]
I think it's unnecessary cast for integer literal.
We can directly define f64 literal like `100_f64`.

Another case is `100_i32 as f64`, I also think it's unnecessary cast.
Same reason as before.
But this will break many test case.

**Comment by gnzlbg** [comment]
Is constructing a floating-point literal exactly the same thing as constructing and integer literal and casting it to a floating point value ? (e.g. are literals truncated according to the same rules as casts?).

**Comment by rink1969** [comment]
I checked the llvm ir generated by rustc.
`100 as f64` `100_f64` and `100_i32 as f64`, these three cases got same irs( a floating-point literal).
`f64::from(100)` has an extra function call.

So I think the cast is unnecessary for integer literal.
But it's ok for var.

**Comment by rink1969** [comment]
I've submit a PR.
https://github.com/rust-lang/rust-clippy/pull/3877
So we can see the effect about the fix.

**Comment by rink1969** [comment]
@gnzlbg  Do you think we can close this issue?

**Comment by flip1995** [comment]
Yes this is implemented in #3877
